File: 100-199/198_HouseRobber.java
// You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security system connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

// Given a list of non-negative integers representing the amount of money of each house, determine the maximum amount of money you can rob tonight without alerting the police.


//Idea: Dynamic Programming, Notation: M[i], the maximum profit after robbering the first i houses. M[i] = max{M[i-1], M[i-2] + nums[i]}

public class Solution {
    public int rob(int[] nums) {
        if(nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        if(nums.length == 2) return Math.max(nums[0],nums[1]);
        int[] max = new int[nums.length];
        for(int i = 0; i < max.length; i++){
            max[i] = 0;
        }
        max[0] = nums[0];
        max[1] = Math.max(nums[0],nums[1]);
        for(int i = 2; i < nums.length; i++){
            max[i] = Math.max(max[i-1],max[i-2] + nums[i]);
        }
        return max[max.length - 1];
    }
}

File: PYTHON/100-199/198_House_Robber.py
#You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security system connected and it will automatically contact the police if two adjacent houses were broken into on the same night.
#
#Given a list of non-negative integers representing the amount of money of each house, determine the maximum amount of money you can rob tonight without alerting the police.
#
#Credits:Special thanks to @ifanchu for adding this problem and creating all test cases. Also thanks to @ts for adding additional test cases.
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0 : return 0
        if len(nums) == 1 : return nums[0]
        if len(nums) == 2 : return max(nums[0],nums[1])
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0],nums[1])
        for i in range(2,len(nums)):
            dp[i] = max(dp[i - 1],dp[i - 2] + nums[i])
        return dp[-1]
        